graphql/executor/extensions.go

```diff
@@ -0,0 +1,147 @@
+package executor
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/99designs/gqlgen/graphql"
+)
+
+// Use adds the given extension to this Executor.
+func (e *Executor) Use(extension graphql.HandlerExtension) {
+	if err := extension.Validate(e.es); err != nil {
+		panic(err)
+	}
+
+	switch extension.(type) {
+	case graphql.OperationParameterMutator,
+		graphql.OperationContextMutator,
+		graphql.OperationInterceptor,
+		graphql.FieldInterceptor,
+		graphql.ResponseInterceptor:
+		e.extensions = append(e.extensions, extension)
+		e.setExtensions()
+
+	default:
+		panic(fmt.Errorf("cannot Use %T as a gqlgen handler extension because it does not implement any extension hooks", extension))
+	}
+}
+
+// AroundFields is a convenience method for creating an extension that only implements field middleware
+func (e *Executor) AroundFields(f graphql.FieldMiddleware) {
+	e.Use(FieldFunc(f))
+}
+
+// AroundOperations is a convenience method for creating an extension that only implements operation middleware
+func (e *Executor) AroundOperations(f graphql.OperationMiddleware) {
+	e.Use(OperationFunc(f))
+}
+
+// AroundResponses is a convenience method for creating an extension that only implements response middleware
+func (e *Executor) AroundResponses(f graphql.ResponseMiddleware) {
+	e.Use(ResponseFunc(f))
+}
+
+func (e *Executor) setExtensions() {
```

Comment:
`setExtensions` is called multiple times, once for each `Use()`. It used to throw the whole executor away and recreate it, but now its reusing it so we need to be careful to zero out all the props.

`operationParameterMutators` and `operationContextMutators` aren't being reset, so if you were to use an extension with it, they would get registered multiple times - once for each other extension thats added after it - and called multiple times at runtime.

Reply:
Ah, the difference is that the old server recreates the private executor (in the [Server constructor](https://github.com/99designs/gqlgen/blob/b165568cce7d5027ba1ee2da68dad3012b12d189/graphql/handler/server.go#L37) and [in `Use()`](https://github.com/99designs/gqlgen/blob/b165568cce7d5027ba1ee2da68dad3012b12d189/graphql/handler/server.go#L78-L90)), ensuring those values are reset properly.

I'll see if I can replicate this behavior in tests and then fix it in this PR.

Thanks for your patience. Judging by your comments, it looks like this PR will take a bit of time to get right :)